crates/ruma-server-util/src/authorization.rs

```diff
@@ -37,204 +43,162 @@ impl XMatrix {
+    }
 }
 
-fn is_alpha(c: u8) -> bool {
-    (0x41..=0x5A).contains(&c) || (0x61..=0x7A).contains(&c)
+impl fmt::Debug for XMatrix {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("XMatrix")
+            .field("origin", &self.origin)
+            .field("destination", &self.destination)
+            .field("key", &self.key)
+            .finish_non_exhaustive()
+    }
 }
 
-fn is_digit(c: u8) -> bool {
-    (0x30..=0x39).contains(&c)
-}
+impl fmt::Display for XMatrix {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let Self { origin, destination, key, sig } = self;
 
-fn is_tchar(c: u8) -> bool {
-    const TOKEN_CHARS: [u8; 15] =
-        [b'!', b'#', b'$', b'%', b'&', b'\'', b'*', b'+', b'-', b'.', b'^', b'_', b'`', b'|', b'~'];
-    is_alpha(c) || is_digit(c) || TOKEN_CHARS.contains(&c)
-}
+        write!(f, r#"{} "#, Self::SCHEME)?;
+
+        if let Some(destination) = destination {
+            write!(f, r#"destination="{destination}","#)?;
+        }
 
-fn is_qdtext(c: u8) -> bool {
-    c == b'\t'
-        || c == b' '
-        || c == 0x21
-        || (0x23..=0x5B).contains(&c)
-        || (0x5D..=0x7E).contains(&c)
-        || is_obs_text(c)
+        write!(f, r#"key="{key}",origin="{origin}",sig="{sig}""#)
+    }
 }
 
-fn is_obs_text(c: u8) -> bool {
-    c >= 0x80 // The spec does contain an upper limit of 0xFF here, but that's enforced by the type
+impl FromStr for XMatrix {
+    type Err = XMatrixParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Self::parse(s)
+    }
 }
 
-fn is_vchar(c: u8) -> bool {
-    (0x21..=0x7E).contains(&c)
+impl TryFrom<&HeaderValue> for XMatrix {
+    type Error = XMatrixParseError;
+
+    fn try_from(value: &HeaderValue) -> Result<Self, Self::Error> {
+        Self::parse(value.to_str()?)
+    }
 }
 
-fn is_quoted_pair(c: u8) -> bool {
-    c == b'\t' || c == b' ' || is_vchar(c) || is_obs_text(c)
+impl TryFrom<&XMatrix> for HeaderValue {
+    type Error = http::header::InvalidHeaderValue;
+
+    fn try_from(value: &XMatrix) -> Result<Self, Self::Error> {
+        value.to_string().try_into()
+    }
 }
 
 impl Credentials for XMatrix {
     const SCHEME: &'static str = "X-Matrix";
 
     fn decode(value: &HeaderValue) -> Option<Self> {
-        let value: Vec<u8> = value.as_bytes().to_vec();
-        parse_xmatrix(&mut value.into_tokens())
+        value.try_into().ok()
     }
 
     fn encode(&self) -> HeaderValue {
-        if let Some(destination) = &self.destination {
-            format!(
-                "X-Matrix origin=\"{}\",destination=\"{destination}\",key=\"{}\",sig=\"{}\"",
-                self.origin, self.key, self.sig
-            )
-        } else {
-            format!("X-Matrix origin=\"{}\",key=\"{}\",sig=\"{}\"", self.origin, self.key, self.sig)
-        }
-        .try_into()
-        .expect("header format is static")
+        self.try_into().expect("header format is static")
```

Comment:
The `sig` value could violate [these conditions](https://docs.rs/http/latest/http/header/struct.HeaderValue.html#method.from_str), should we make a new type for it so it can be validated?

Reply:
Example:
```rust
    #[test]
    fn xmatrix_auth_1_3_extended_ascii() {
        let header = HeaderValue::from_static("X-Matrix origin=\"origin.hs.example.com\"  ,     destination=\"destination.hs.example.com\",key=\"ed25519:key1\", sig=\"ABCDEFÇ...\"");
        let credentials = XMatrix::try_from(&header).unwrap();

        assert_eq!(credentials.origin, "origin.hs.example.com");
        assert_eq!(credentials.destination.unwrap(), "destination.hs.example.com");
        assert_eq!(credentials.key, "ed25519:key1");
        assert_eq!(credentials.sig, "ABCDEFÇ...");
    }
```

```
---- authorization::tests::xmatrix_auth_1_3_extended_ascii stdout ----
thread 'authorization::tests::xmatrix_auth_1_3_extended_ascii' panicked at /home/alpha/.local/share/cargo/registry/src/index.crates.io-6f17d22bba15001f/http-1.1.0/src/header/value.rs:95:17:
index out of bounds: the len is 0 but the index is 0
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    authorization::tests::xmatrix_auth_1_3_extended_ascii
```

Reply:
Hmm, I'm not sure anymore, since the following passes fine:
```rust
    #[test]
    fn xmatrix_auth_1_3_extended_ascii() {
        let creds = XMatrix {
            origin: "origin.hs.example.com".try_into().unwrap(),
            destination: Some("destination.hs.example.com".try_into().unwrap()),
            key: "ed25519:key1".try_into().unwrap(),
            sig: "ABCDEFÇ...".to_owned(),
        };

        let _header = creds.encode();
    }
```

Reply:
Seems like it is escaped: 
```
"X-Matrix destination=\"destination.hs.example.com\",key=\"ed25519:key1\",origin=\"origin.hs.example.com\",sig=\"ABCDEF\xc3\x87...\""
```

So is there a reason to have an `expect()` in `decode` instead of just moving it to `From`?

Reply:
I don't see any evidence of escaping there, that's just `String`'s `Debug` implementation.